step-framework-parent/step-framework-timeseries/src/main/java/step/core/timeseries/aggregation/TimeSeriesAggregationPipeline.java

```diff
@@ -5,60 +5,169 @@
+        }
 
-    public TimeSeriesAggregationPipeline(Collection<Bucket> collectionDriver, long resolution) {
-        this.collection = collectionDriver;
-        this.sourceResolution = resolution;
+        long resultResolution = sourceResolution;
+        Long resultFrom = roundDownToMultiple(query.getFrom(), sourceResolution);
+        Long resultTo = roundUpToMultiple(query.getTo(), sourceResolution);
+        if (query.isShrink()) { // we expand the interval to the closest completed resolutions
+            resultResolution = resultTo - resultFrom;
+        } else {
+            Integer bucketsCount = query.getBucketsCount();
+            if (bucketsCount != null && bucketsCount > 0) {
+                if ((resultTo - resultFrom) / sourceResolution <= bucketsCount) { // not enough buckets
+                    resultResolution = sourceResolution;
+                } else {
+                    long difference = resultTo - resultFrom;
+                    resultResolution = Math.round(difference / (double) bucketsCount);
+                    // there are situation when resultResolution/sourceResolution is below 0.5, and that would end up rounded in 0.
+                    resultResolution = Math.max(Math.round((double) resultResolution / sourceResolution), 1) * sourceResolution; // round to nearest multiple, up or down
+                }
+            } else {
+                Long proposedResolution = query.getBucketsResolution();
+                if (proposedResolution != null && proposedResolution != 0) {
+                    resultResolution = Math.max(sourceResolution, roundDownToMultiple(proposedResolution, sourceResolution));
+                    resultResolution = roundDownToMultiple(resultResolution, sourceResolution);
+                    long diff = resultTo - resultFrom;
+                    diff = roundUpToMultiple(diff, resultResolution);
+                    resultTo = resultFrom + diff;
+                }
+            }
+        }
+        return new TimeSeriesProcessedParams()
+                .setFrom(resultFrom)
+                .setTo(resultTo)
+                .setResolution(resultResolution)
+                .setGroupDimensions(query.getGroupDimensions())
+                .setFilter(query.getFilter())
+                .setShrink(query.isShrink())
+                .setCollectAttributeKeys(query.getCollectAttributeKeys())
+                .setCollectAttributesValuesLimit(query.getCollectAttributesValuesLimit());
+    }
+    
+    private static long roundUpToMultiple(long value, long multiple) {
+        return (long) Math.ceil((double) value / multiple) * multiple;
     }
 
-    protected long getSourceResolution() {
-        return sourceResolution;
+    private static long roundDownToMultiple(long value, long multiple) {
+        return value - value % multiple;
+    }
+    
+    private TimeSeriesCollection chooseTargetCollection(TimeSeriesAggregationQuery query) {
```

Comment:
these set of methods should be refactored to be more efficient, there is no need to have the 3 different data structured to store the same thing, all are used while choosing the target collection. Also we go through them multiple times and in different ways to find the existing resolution and then checking if data still exists for it which can and should ne avoided

Reply:
after the changes, the only remaining one is resolutionIndexes. This is just to directly access the collections by indexes. If you prefer we can also iterate each time.